subsys/bluetooth/controller/multithreading_lock.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @lock.h
+ *
+ * @brief This file defines APIs needed to lock the BLE controller APIs for threadsafe operation.
+ */
+
+#ifndef LOCK_H__
+#define LOCK_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr.h>
+#include <soc.h>
+#include <gpio.h>
+
+/** Macro representation for unlock API */
+#define API_UNLOCK api_unlock
+
+/** Macro representation for typical usage of locking API */
+#define API_LOCK_AND_RETURN_ON_FAIL(err) \
+	do {				 \
+		err = api_lock();	 \
+		if (err != 0) {		 \
+			return err;	 \
+		}			 \
+	} while (0);
+
```

Comment:
What about the following:
```
#ifdef BLECTLR_CONFIG_THREADSAFE
#define THREADSAFE_CALL(api_call) \
 do {
    api_lock();
    api_call();
    api_unlock();
  while (0);
#else
THREADSAFE_CALL(api_call) api_call
#endif
```
(error code return is not included in this example)

Reply:
That will not work. Most api_calls() have return values and depending upon the returns driver code will work differently.

Reply:
The error code can be wrapped as well, can't it?

Reply:
I think you should look at this once more